src/locks.rs

```diff
@@ -0,0 +1,146 @@
+//! Lock strategies.
+
+#[cfg(unix)]
+use std::os::fd::AsRawFd;
+use std::{
+    fs::File,
+    ops::{Deref, DerefMut},
+};
+
+use memmap2::MmapMut;
+
+use crate::synchronizer::SynchronizerError;
+
+/// The write lock strategy supports different lock implementations which can be chosen based on
+/// the guarantees required, platform support, and performance constraints.
+pub trait WriteLockStrategy<'a> {
+    type Guard: DerefMut<Target = MmapMut> + 'a;
+
+    /// Create a new instance of this lock strategy.
+    ///
+    /// The `mmap` parameter will have write access controlled by the lock.
+    ///
+    /// The `file` parameter is required because lock strategies depending on `flock` must hold
+    /// the file descriptor open so the kernel does not release the lock.
+    fn new(mmap: MmapMut, file: File) -> Self;
+
+    /// Provide read access to mmaped memory.
```

Comment:
If it's read access only, then maybe use `Mmap` and not `MmapMut`? It can be converted back to mutable with:
https://docs.rs/memmap2/latest/memmap2/struct.Mmap.html#method.make_mut

Reply:
We are already returning `&Mmap` instead of `Mmap`, which prevents mutating the data (i.e., via `Deref<Target=[u8]>`. Technically, `&MmapMut` does have a few methods that `&Mmap` does not, but these appear related to the permissions attached to the underlying file descriptor and are orthogonal to data mutability.

Calls to `.make_mut()`/`.make_read_only()` result in a call to `libc::mprotect`, which incurs some performance penalty and would introduce an error case, complicating the API.

If you prefer, we could return `&[u8]` or a newtype to avoid exposing types from the `memmap2` crate.

Reply:
Returning `&[u8]` sounds reasonable.

Reply:
Updated to return `&[u8]`.